cpp/path_module/algorithm/path.cpp

```diff
@@ -1,301 +1,343 @@
-  for (auto rel : rels) {
-    // go through every relationship of the node and expand to the other node of the relationship
-    if (relationships_set.find(rel) !=
-        relationships_set.end()) {  // relationships_set contains all relationships already visited in this path, and
-                                    // the usage of this if loop is to evade cycles
-      continue;
+void Path::Create(mgp_list *args, mgp_graph *memgraph_graph, mgp_result *result, mgp_memory *memory) {
+  mgp::MemoryDispatcherGuard guard{memory};
+  const auto arguments = mgp::List(args);
+  const auto record_factory = mgp::RecordFactory(result);
+  try {
+    auto start_node{arguments[0].ValueNode()};
+    auto relationships{arguments[1].ValueMap()};
+
+    mgp::Path path{start_node};
+    for (const auto &relationship : relationships["rel"].ValueList()) {
+      if (relationship.IsNull()) {
+        break;
+      }
+      if (!relationship.IsRelationship()) {
+        std::ostringstream oss;
+        oss << relationship.Type();
+        throw mgp::ValueException("Expected relationship or null type, got " + oss.str());
+      }
+
+      const auto rel = relationship.ValueRelationship();
+      auto last_node = path.GetNodeAt(path.Length());
+
+      bool endpoint_is_from = false;
+
+      if (last_node.Id() == rel.From().Id()) {
+        endpoint_is_from = true;
+      }
+
+      auto contains = [](mgp::Relationships relationships, const mgp::Id id) {
+        for (const auto relationship : relationships) {
+          if (relationship.To().Id() == id) {
+            return true;
+          }
+        }
+        return false;
+      };
+
+      if ((endpoint_is_from && !contains(rel.From().OutRelationships(), rel.To().Id())) ||
+          (!endpoint_is_from && !contains(rel.To().OutRelationships(), rel.From().Id()))) {
+        break;
+      }
+
+      path.Expand(rel);
     }
-    mgp::Path cpy = mgp::Path(path);
-    const std::string rel_type = std::string(rel.Type());
-    bool rel_allowed = outgoing ? RelationshipAllowed(rel_type, relationshipSets, any_outgoing, any_incoming, true)
-                                : RelationshipAllowed(rel_type, relationshipSets, any_outgoing, any_incoming, false);
-    if (!rel_allowed) {  // if relationship not allowed, go to next one
+
+    auto record = record_factory.NewRecord();
+    record.Insert(std::string(kResultCreate).c_str(), path);
+
+  } catch (const std::exception &e) {
+    record_factory.SetErrorMessage(e.what());
+    return;
+  }
+}
+
+void Path::PathExpand::ExpandPath(mgp::Path &path, const mgp::Relationship &relationship, int64_t path_size) {
+  path.Expand(relationship);
+  visited_.insert(relationship.Id().AsInt());
+  DFS(path, path_size + 1);
+  visited_.erase(relationship.Id().AsInt());
+  path.Pop();
+}
+
+void Path::PathExpand::Expand(mgp::Path &path, mgp::Relationships relationships, bool outgoing, int64_t path_size,
```

Comment:
Can this part have different naming than mgp::Path::Exand, it is difficult to track. Maybe  `StartExpanding`

Reply:
And this also has same name as Path::Exapnd, so we have three different expands, it is quite a lot of similar names.  mgp::Path::Expand stays as is, Path::Expand can be called Path::ExpandProc, and Path::PathExpand::Expand can be call Path::PathExpand::StartExpanding

Reply:
Path::PathExpand::Expand -> ExpandFromRelationships
I think other can stay the same since Path::Expand is called that because the procedure is called path.expand